lib/web/apps.go

```diff
@@ -0,0 +1,387 @@
+	// is running.
+	ClusterName string
+}
+
+// resolveDirect takes a public address and cluster name and exactly resolves
+// the application and the server on which it is running.
+func (h *Handler) resolveDirect(ctx context.Context, publicAddr string, clusterName string) (*services.App, services.Server, string, error) {
+	clusterClient, err := h.cfg.Proxy.GetSite(clusterName)
+	if err != nil {
+		return nil, nil, "", trace.Wrap(err)
+	}
+
+	authClient, err := clusterClient.GetClient()
+	if err != nil {
+		return nil, nil, "", trace.Wrap(err)
+	}
+
+	app, server, err := h.match(ctx, authClient, matchPublicAddr(publicAddr))
+	if err != nil {
+		return nil, nil, "", trace.Wrap(err)
+	}
+
+	return app, server, clusterName, nil
+}
+
+// resolveFQDN makes a best effort attempt to resolve FQDN to an application
+// running a root or leaf cluster.
+//
+// Known edge cases this function can incorrectly resolve an application exist.
+// For example, if you have an application named "acme" within both the root
+// and leaf cluster, this method will always return "acme" running within the
+// root cluster. Always supply public address and cluster name to
+// deterministically resolve an application.
+func (h *Handler) resolveFQDN(ctx context.Context, fqdn string) (*services.App, services.Server, string, error) {
+	// Parse the address to remove the port if it's set.
+	addr, err := utils.ParseAddr(fqdn)
+	if err != nil {
+		return nil, nil, "", trace.Wrap(err)
+	}
+
+	// Try and match FQDN to public address of application within cluster.
+	app, server, err := h.match(ctx, h.cfg.ProxyClient, matchPublicAddr(addr.Host()))
+	if err == nil {
+		return app, server, h.auth.clusterName, nil
+	}
+
+	// Extract the first subdomain from the FQDN and attempt to use this as the
+	// application name.
+	appName := strings.Split(addr.Host(), ".")[0]
+
+	// Try and match application name to an application within the cluster.
+	app, server, err = h.match(ctx, h.cfg.ProxyClient, matchName(appName))
+	if err == nil {
+		return app, server, h.auth.clusterName, nil
+	}
+
+	// Loop over all clusters and try and match application name to an
+	// application with the cluster.
+	for _, remoteClient := range h.cfg.Proxy.GetSites() {
+		authClient, err := remoteClient.CachingAccessPoint()
+		if err != nil {
+			return nil, nil, "", trace.Wrap(err)
+		}
+
+		app, server, err = h.match(ctx, authClient, matchName(appName))
+		if err == nil {
+			return app, server, remoteClient.GetName(), nil
+		}
+	}
+
+	return nil, nil, "", trace.NotFound("failed to resolve %v to any application within any cluster", fqdn)
+}
+
+// match will match an application with the passed in matcher function. Matcher
+// functions that can match on public address and name are available.
+func (h *Handler) match(ctx context.Context, authClient appGetter, fn matcher) (*services.App, services.Server, error) {
+	servers, err := authClient.GetAppServers(ctx, defaults.Namespace)
```

Comment:
Seems like the same logic as `getApp()` method in `session.go`, possible to reuse it?